File: tests/test_decompressor.py
import unittest
from io import BytesIO

try:
    import micropython
except ImportError:
    micropython = None

Decompressors = []
decompresses = []

if micropython is None:
    from tamp.decompressor import Decompressor as PyDecompressor
    from tamp.decompressor import decompress as py_decompress

    Decompressors.append(PyDecompressor)
    decompresses.append(py_decompress)

    try:
        from tamp._c_decompressor import Decompressor as CDecompressor
        from tamp._c_decompressor import decompress as c_decompress

        Decompressors.append(CDecompressor)
        decompresses.append(c_decompress)
    except ImportError:
        pass

else:
    try:
        from tamp_native import Decompressor as NativeDecompressor
        from tamp_native import decompress as native_decompress

        Decompressors.append(NativeDecompressor)
        decompresses.append(native_decompress)
    except ImportError:
        pass


class TestDecompressor(unittest.TestCase):
    def test_decompressor_basic(self):
        for Decompressor in Decompressors:
            with self.subTest(Decompressor=Decompressor):
                expected = b"foo foo foo"

                compressed = bytes(
                    # fmt: off
                    [
                        0b010_11_00_0,  # header (window_bits=10, literal_bits=8)
                        0b1_0110011,  # literal "f"
                        0b0_0_0_00100,  # the pre-init buffer contains "oo" at index 131
                        # size=2 -> 0b0
                        # 131 -> 0b0010000011
                        0b00011_1_00,  # literal " "
                        0b100000_0_1,  # There is now "foo " at index 0
                        0b000_00000,  # size=4 -> 0b1000
                        0b00000_0_11,  # Just "foo" at index 0; size=3 -> 0b11
                        0b00000000,  # index 0 -> 0b0000000000
                        0b00_000000,  # 6 bits of zero-padding
                    ]
                    # fmt: on
                )

                with BytesIO(compressed) as f:
                    decompressor = Decompressor(f)
                    actual = decompressor.read()

                self.assertEqual(actual, expected)

    def test_decompressor_restricted_read_size(self):
        for Decompressor in Decompressors:
            with self.subTest(Decompressor=Decompressor):
                compressed = bytes(
                    # fmt: off
                    [
                        0b010_11_00_0,  # header (window_bits=10, literal_bits=8)
                        0b1_0110011,  # literal "f"
                        0b0_0_0_00100,  # the pre-init buffer contains "oo" at index 131
                        # size=2 -> 0b0
                        # 131 -> 0b0010000011
                        0b00011_1_00,  # literal " "
                        0b100000_0_1,  # There is now "foo " at index 0
                        0b000_00000,  # size=4 -> 0b1000
                        0b00000_0_11,  # Just "foo" at index 0; size=3 -> 0b11
                        0b00000000,  # index 0 -> 0b0000000000
                        0b00_000000,  # 6 bits of zero-padding
                    ]
                    # fmt: on
                )

                with BytesIO(compressed) as f:
                    decompressor = Decompressor(f)
                    self.assertEqual(decompressor.read(4), b"foo ")
                    self.assertEqual(decompressor.read(2), b"fo")
                    self.assertEqual(decompressor.read(-1), b"o foo")

    def test_decompressor_flushing(self):
        for decompress in decompresses:
            with self.subTest(decompress=decompress):
                compressed = bytes(
                    # fmt: off
                    [
                        0b010_11_00_0,  # header (window_bits=10, literal_bits=8)
                        0b1_0101000,  # literal 'Q'
                        0b1_0_101010,  # FLUSH_CODE
                        0b11_000000,  # FLUSH CONTINUE
                        0b1_0101011,  # literal 'W'
                        0b1_0_101010,  # FLUSH_CODE
                        0b11_000000,  # FLUSH CONTINUE
                    ]
                    # fmt: on
                )
                decoded = decompress(compressed)
                self.assertEqual(decoded, b"QW")

    def test_decompressor_missing_dict(self):
        for Decompressor in Decompressors:
            # fmt: off
            with self.subTest(Decompressor=Decompressor), \
                 self.assertRaises(ValueError), \
                 BytesIO(bytes([0b000_10_1_0_0])) as f:
                Decompressor(f)
            # fmt: on

    def test_decompressor_full_output_dst_immediately_after_src(self):
        # Decompressor's perspective of window
        # compressed data: b"z"
        #    * 0 write literal "a"   -> b"abcd"
        #    * 1 write pattern "abc"
        custom_dictionary = bytearray(1024)
        custom_dictionary_init = b"abcd"
        custom_dictionary[: len(custom_dictionary_init)] = custom_dictionary_init

        data = bytes(
            [
                # fmt: off
                # header (window_bits=10, literal_bits=8, custom)
                0b010_11_1_0_0,
                0b1_0110000,  # literal "a"
                0b1_0_11_0000,  # token "abc"
                0b000000_00,
                # 2-bit padding
                # fmt: on
            ]
        )

        for Decompressor in Decompressors:
            with self.subTest(Decompressor=Decompressor):
                with BytesIO(data) as f:
                    # Sanity check that without limiting output, it decompresses correctly.
                    decompressor = Decompressor(f, dictionary=bytearray(custom_dictionary))
                    self.assertEqual(decompressor.read(), b"aabc")

                with BytesIO(data) as f:
                    decompressor = Decompressor(f, dictionary=bytearray(custom_dictionary))
                    self.assertEqual(decompressor.read(1), b"a")
                    self.assertEqual(decompressor.read(1), b"a")
                    self.assertEqual(decompressor.read(1), b"b")
                    self.assertEqual(decompressor.read(1), b"c")

    def test_decompressor_partial_token(self):
        compressed = bytes(
            # fmt: off
            [
                0b010_11_00_0,  # header (window_bits=10, literal_bits=8)
                0b1_0110011,  # literal "f"
                0b0_0_0_00100,  # the pre-init buffer contains "oo" at index 131
                # size=2 -> 0b0
                # 131 -> 0b0010000011
                0b00011_1_00,  # literal " "
                0b100000_0_1,  # There is now "foo " at index 0
                0b000_00000,  # size=4 -> 0b1000
                ####################### - stream-break here
                0b00000_0_11,  # Just "foo" at index 0; size=3 -> 0b11
                0b00000000,  # index 0 -> 0b0000000000
                0b00_000000,  # 6 bits of zero-padding
            ]
            # fmt: on
        )

        expected = b"foo foo foo"

        for Decompressor in Decompressors:
            with self.subTest(Decompressor=Decompressor), BytesIO(compressed[:6]) as f:
                decompressor = Decompressor(f)
                read0 = decompressor.read()

                f.write(compressed[6:])
                f.seek(6)

                read1 = decompressor.read()
                self.assertEqual(read0 + read1, expected)

    def test_decompressor_extended_rle(self):
        """Decompress extended-format data containing an RLE token."""
        for Decompressor in Decompressors:
            with self.subTest(Decompressor=Decompressor):
                # 20 bytes of 'A': literal 'A' + RLE(count=19)
                compressed = bytes(
                    # fmt: off
                    [
                        0x5A,  # header: 010_11_0_1_0 (window=10, literal=8, extended=1)
                        0xA0,  # literal 'A' bits + start token flag
                        0xAA,  # RLE huffman sym12
                        0xB1,  # count upper=sym1, lower=0001, pad
                    ]
                    # fmt: on
                )
                with BytesIO(compressed) as f:
                    decompressor = Decompressor(f)
                    actual = decompressor.read()
                self.assertEqual(actual, b"A" * 20)

    def test_decompressor_extended_rle_short(self):
        """Decompress extended RLE with small count (count=4)."""
        for Decompressor in Decompressors:
            with self.subTest(Decompressor=Decompressor):
                compressed = bytes(
                    # fmt: off
                    [
                        0x5A,  # header: 010_11_0_1_0 (window=10, literal=8, extended=1)
                        0xA1,  # literal 'B' bits + start token flag
                        0x2A,  # ...0 from B, token flag, RLE huffman partial
                        0x84,  # ...huffman end, count upper=sym0, lower=0010, pad
                    ]
                    # fmt: on
                )
                with BytesIO(compressed) as f:
                    decompressor = Decompressor(f)
                    actual = decompressor.read()
                self.assertEqual(actual, b"B" * 5)

    def test_decompressor_extended_match(self):
        """Decompress extended-format data containing an extended match token."""
        for Decompressor in Decompressors:
            with self.subTest(Decompressor=Decompressor):
                # 14-byte match at index 0 in custom dictionary
                pattern = b"abcdefghijklmn"
                dictionary = bytearray(1 << 8)
                dictionary[: len(pattern)] = pattern

                compressed = bytes(
                    # fmt: off
                    [
                        0x1E,  # header: 000_11_1_1_0 (window=8, literal=8, custom=1, extended=1)
                        0x4E,  # ext_match token + count upper=sym0
                        0x00,  # count lower=000, position=0 partial
                        0x00,  # position end, padding
                    ]
                    # fmt: on
                )
                with BytesIO(compressed) as f:
                    decompressor = Decompressor(f, dictionary=dictionary)
                    actual = decompressor.read()
                self.assertEqual(actual, pattern)

    def test_decompressor_extended_rle_restricted_read(self):
        """Decompress RLE data with restricted read sizes to test partial output."""
        for Decompressor in Decompressors:
            with self.subTest(Decompressor=Decompressor):
                compressed = bytes(
                    # fmt: off
                    [
                        0x5A,  # header: 010_11_0_1_0 (window=10, literal=8, extended=1)
                        0xA0,  # literal 'A' bits + start token flag
                        0xAA,  # RLE huffman sym12
                        0xB1,  # count upper=sym1, lower=0001, pad
                    ]
                    # fmt: on
                )
                with BytesIO(compressed) as f:
                    decompressor = Decompressor(f)
                    part1 = decompressor.read(5)
                    part2 = decompressor.read(10)
                    part3 = decompressor.read(-1)
                self.assertEqual(part1 + part2 + part3, b"A" * 20)

    def test_decompressor_readinto(self):
        for Decompressor in Decompressors:
            with self.subTest(Decompressor=Decompressor):
                expected = b"foo foo foo"

                compressed = bytes(
                    # fmt: off
                    [
                        0b010_11_00_0,  # header (window_bits=10, literal_bits=8)
                        0b1_0110011,  # literal "f"
                        0b0_0_0_00100,  # the pre-init buffer contains "oo" at index 131
                        # size=2 -> 0b0
                        # 131 -> 0b0010000011
                        0b00011_1_00,  # literal " "
                        0b100000_0_1,  # There is now "foo " at index 0
                        0b000_00000,  # size=4 -> 0b1000
                        0b00000_0_11,  # Just "foo" at index 0; size=3 -> 0b11
                        0b00000000,  # index 0 -> 0b0000000000
                        0b00_000000,  # 6 bits of zero-padding
                    ]
                    # fmt: on
                )

                actual_buf = bytearray(1024)
                with BytesIO(compressed) as f:
                    decompressor = Decompressor(f)
                    actual_len = decompressor.readinto(actual_buf)

                self.assertEqual(actual_len, len(expected))
                self.assertEqual(actual_buf[:actual_len], expected)

File: ctests/test_decompressor.c
#include <assert.h>
#include <stdbool.h>
#include <stddef.h>
#include <string.h>

#include "tamp/compressor.h"
#include "tamp/decompressor.h"
#include "unity.h"

void test_decompressor_byte_by_byte(void) {
    /*****
     * Tests the decompressor if we feed in 1 byte at a time.
     */
    const unsigned char compressed[] = {
        0b01011000,  // header (window_bits=10, literal_bits=8)
        0b10110011,  // literal "f"
        0b00000100,  // the pre-init buffer contains "oo" at index 131
                     // size=2 -> 0b0
                     // 131 -> 0b0010000011
        0b00011100,  // literal " "
        0b10000001,  // There is now "foo " at index 0
        0b00000000,  // size=4 -> 0b1000
        0b00000011,  // Just "foo" at index 0; size=3 -> 0b11
        0b00000000,  // index 0 -> 0b0000000000
        0b00000000,  // 6 bits of zero-padding
    };

    tamp_res res;
    TampDecompressor d;
    unsigned char window_buffer[1 << 10];

    res = tamp_decompressor_init(&d, NULL, window_buffer, 10);
    TEST_ASSERT_EQUAL(res, TAMP_OK);

    unsigned char input_byte;
    size_t read_size = 0;
    size_t input_chunk_consumed_size;
    bool file_exhausted = false;

    size_t output_written_size = 0;

    unsigned char output_buffer_complete[32] = {0};
    unsigned char *output_buffer = output_buffer_complete;

    size_t output_chunk_written_size;

    int j = 0;
    while (true) {
        res = tamp_decompressor_decompress(&d, output_buffer, sizeof(output_buffer), &output_chunk_written_size,
                                           &input_byte, read_size, &input_chunk_consumed_size);
        assert(res >= TAMP_OK);  // i.e. an "ok" result
        assert(input_chunk_consumed_size == read_size);

        output_written_size += output_chunk_written_size;
        output_buffer += output_chunk_written_size;

        if (res == TAMP_OUTPUT_FULL) break;
        if (res == TAMP_INPUT_EXHAUSTED && file_exhausted) break;

        if (j >= sizeof(compressed)) {
            read_size = 0;
        } else {
            input_byte = compressed[j];
            j++;
            read_size = 1;
        }

        file_exhausted = read_size == 0;
    }

    TEST_ASSERT_EQUAL_STRING("foo foo foo", output_buffer_complete);
}

void test_decompressor_malicious_oob(void) {
    /*****
     * Tests the decompressor if we feed in a pattern at position WINDOW_SIZE - 1.
     * The compressor should never generate this.
     */
    const unsigned char compressed[] = {
        0b01011000,  // header (window_bits=10, literal_bits=8)
        0b00111111,  // pattern of length 2 at WINDOW_SIZE -1
        0b11110000,  // 4 bits of zero-padding
    };

    tamp_res res;
    TampDecompressor d;
    unsigned char window_buffer[1 << 10];

    res = tamp_decompressor_init(&d, NULL, window_buffer, 10);
    TEST_ASSERT_EQUAL(res, TAMP_OK);

    unsigned char output_buffer[32];
    size_t output_written_size;

    res = tamp_decompressor_decompress(&d, output_buffer, sizeof(output_buffer), &output_written_size, compressed,
                                       sizeof(compressed), NULL);
    TEST_ASSERT_EQUAL(res, TAMP_OOB);
}

void test_decompressor_extended_rle(void) {
    /*****
     * Tests decompression of extended-format data with an RLE token.
     * Compressed data encodes 20 bytes of 'A' (literal 'A' + RLE count=19).
     */
    const unsigned char compressed[] = {
        0x5A,  // header: window=10, literal=8, extended=1
        0xA0,  // literal 'A' + start of RLE token
        0xAA,  // RLE huffman continued
        0xB1,  // count encoding + padding
    };

    tamp_res res;
    TampDecompressor d;
    unsigned char window_buffer[1 << 10];

    res = tamp_decompressor_init(&d, NULL, window_buffer, 10);
    TEST_ASSERT_EQUAL(TAMP_OK, res);

    unsigned char output[32];
    size_t output_written;

    res =
        tamp_decompressor_decompress(&d, output, sizeof(output), &output_written, compressed, sizeof(compressed), NULL);
    TEST_ASSERT_GREATER_OR_EQUAL(TAMP_OK, res);
    TEST_ASSERT_EQUAL(20, output_written);

    // Verify all bytes are 'A'
    unsigned char expected[20];
    memset(expected, 'A', sizeof(expected));
    TEST_ASSERT_EQUAL_MEMORY(expected, output, 20);
}

void test_decompressor_extended_match(void) {
    /*****
     * Tests decompression of extended-format data with an extended match token.
     * Compressed data: 14-byte match at index 0 using custom dictionary.
     * conf=NULL so header is read from the compressed data.
     */
    const unsigned char compressed[] = {
        0x1E,  // header: window=8, literal=8, custom=1, extended=1
        0x4E,  // ext_match token + count encoding
        0x00,  // position + padding
        0x00,  // padding
    };

    // Pre-initialize the window buffer with the dictionary
    unsigned char window[1 << 8];
    memset(window, 0, sizeof(window));
    const char *pattern = "abcdefghijklmn";
    memcpy(window, pattern, 14);

    tamp_res res;
    TampDecompressor d;

    // conf=NULL: header is read from compressed data during decompress.
    // Since header has custom_dict=1, the window is preserved (not re-initialized).
    res = tamp_decompressor_init(&d, NULL, window, 8);
    TEST_ASSERT_EQUAL(TAMP_OK, res);

    unsigned char output[32];
    size_t output_written;

    res =
        tamp_decompressor_decompress(&d, output, sizeof(output), &output_written, compressed, sizeof(compressed), NULL);
    TEST_ASSERT_GREATER_OR_EQUAL(TAMP_OK, res);
    TEST_ASSERT_EQUAL(14, output_written);
    TEST_ASSERT_EQUAL_MEMORY(pattern, output, 14);
}

void test_decompress_cb_callback_receives_input_consumed(void) {
    // First, compress some data to get valid compressed input
    TampCompressor compressor;
    unsigned char window[1 << 10];
    unsigned char compressed[256];
    size_t compressed_size, input_consumed;
    TampConf conf = {.window = 10, .literal = 8};

    tamp_compressor_init(&compressor, &conf, window);

    const char *original =
        "The quick brown fox jumps over the lazy dog. "
        "The quick brown fox jumps over the lazy dog.";
    size_t original_len = strlen(original);

    tamp_compressor_compress_and_flush(&compressor, compressed, sizeof(compressed), &compressed_size,
                                       (const unsigned char *)original, original_len, &input_consumed, false);

    // Now decompress with callback tracking
    // Skip the header byte for total_bytes calculation
    TampDecompressor decompressor;
    tamp_decompressor_init(&decompressor, NULL, window, 10);

    unsigned char output[256];
    size_t output_written;
    size_t decompress_input_consumed;

    CallbackTracker tracker;
    callback_tracker_init(&tracker, compressed_size);

    tamp_res res =
        tamp_decompressor_decompress_cb(&decompressor, output, sizeof(output), &output_written, compressed,
                                        compressed_size, &decompress_input_consumed, tracking_callback, &tracker);

    // Callback should have been called
    TEST_ASSERT_GREATER_THAN(0, tracker.call_count);

    // Monotonicity and total_bytes consistency are checked in the callback itself.
    // Final bytes_processed should equal input_consumed
    TEST_ASSERT_EQUAL(decompress_input_consumed, tracker.last_bytes_processed);

    // Verify decompression worked
    TEST_ASSERT_EQUAL(original_len, output_written);
    TEST_ASSERT_EQUAL_MEMORY(original, output, original_len);
}
